# DAAD Data Cleaning

In this notebook we are going to clean the data set we grab from the DAAD webpage and let it ready to analyze later.

## Imports


In [1]:
import warnings
import re
import pandas as pd
warnings.filterwarnings('ignore')
#pd.set_option("display.max_rows", None, "display.max_columns", None)

In [2]:
# Jason Load/Save files
# Load Data

import json
def load_data(title):
    with open(title, encoding='utf-8') as f:
        return json.load(f)


## Pre-Analysis

In [3]:
# Create DF
courses = load_data('./Database/daad_courses_final_1.json')

df = pd.DataFrame(courses)


In [4]:
df.describe(include='all')


Couse ID                                            Program  \
count   2047.000000                                               2047   
unique          NaN                                               1961   
top             NaN  Master of Science in Economics Master of Scien...   
freq            NaN                                                 10   
mean    5036.176356                                                NaN   
std     1062.226761                                                NaN   
min     3589.000000                                                NaN   
25%     4160.500000                                                NaN   
50%     4772.000000                                                NaN   
75%     5696.500000                                                NaN   
max     7169.000000                                                NaN   

                                 University             Degree  \
count                                  2047               1705   
unique                                  289                773   
top     University of Göttingen • Göttingen  Master of Science   
freq                                     71                320   
mean                                    NaN                NaN   
std                                     NaN                NaN   
min                                     NaN                NaN   
25%                                     NaN                NaN   
50%                                     NaN                NaN   
75%                                     NaN                NaN   
max                                     NaN                NaN   

       Teaching language Languages Programme duration        Beginning  \
count               2002      1729               1705             1705   
unique                37       986                  8                5   
top              English   English        4 semesters  Winter semester   
freq                1523       103                993             1015   
mean                 NaN       NaN                NaN              NaN   
std                  NaN       NaN                NaN              NaN   
min                  NaN       NaN                NaN              NaN   
25%                  NaN       NaN                NaN              NaN   
50%                  NaN       NaN                NaN              NaN   
75%                  NaN       NaN                NaN              NaN   
max                  NaN       NaN                NaN              NaN   

       More information on beginning of studies  \
count                                       782   
unique                                      529   
top                                     October   
freq                                         47   
mean                                        NaN   
std                                         NaN   
min                                         NaN   
25%                                         NaN   
50%                                         NaN   
75%                                         NaN   
max                                         NaN   

                             Application deadline  ...  \
count                                        1638  ...   
unique                                       1142  ...   
top     15 July for the following winter semester  ...   
freq                                           44  ...   
mean                                          NaN  ...   
std                                           NaN  ...   
min                                           NaN  ...   
25%                                           NaN  ...   
50%                                           NaN  ...   
75%                                           NaN  ...   
max                                           NaN  ...   

       Phase(s) of compulsory attendance  \
count                                  6   
unique                                 2  

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2047 entries, 0 to 2046
Columns: 104 entries, Couse ID to Course language
dtypes: int64(1), object(103)
memory usage: 1.6+ MB


There are 104 Features, thus we need to select the most relevante for us

In [6]:
col_features = ['Couse ID','University', 'Program', 'Degree', 'Teaching language','Languages','Programme duration','Beginning','Application deadline',
               'Tuition fees per semester in EUR','Description/content','Types of assessment', 'Integrated internships','Course-specific, integrated German language courses','Semester contribution',
               'Academic admission requirements','Language requirements','Submit application to','Accommodation','Possibility of finding part-time employment']

In [7]:
# The new data frame is called programmes
programmes = df[col_features].copy()

In [8]:
programmes.head()

Couse ID                                 University  \
0      4000  Friedrich Schiller University Jena • Jena   
1      4001          University of Augsburg • Augsburg   
2      4002     Philipps-Universität Marburg • Marburg   
3      4003        University of Stuttgart • Stuttgart   
4      4004               University of Cologne • Köln   

                                             Program  \
0  Master of Science in Molecular Medicine MSc Mo...   
1  Master of Science in Software Engineering Mast...   
2  German as a Foreign Language (MA) German as a ...   
3  Graduate School Simulation Technology (GS SimT...   
4  Bonn-Cologne Graduate School of Physics and As...   

                                      Degree Teaching language  \
0    Master of Science in Molecular Medicine           English   
1  Master of Science in Software Engineering           English   
2                             Master of Arts            German   
3                                        PhD    German English   
4                 Doctoral Degree in Physics           English   

                                           Languages Programme duration  \
0                       Courses are held in English.        4 semesters   
1  The courses are in English unless all students...        4 semesters   
2      German in courses and for the Master's thesis        4 semesters   
3                English about 60%, German about 40%        6 semesters   
4         The entire programme is taught in English.        6 semesters   

                                Beginning  \
0                         Winter semester   
1                         Winter semester   
2                         Winter semester   
3  Only for doctoral programmes: any time   
4  Only for doctoral programmes: any time   

                                Application deadline  \
0  31 May for the following winter semester, appl...   
1                          1 May, for all applicants   
2          15 July for the following winter semester   
3                            No application deadline   
4                                       No deadlines   

  Tuition fees per semester in EUR  \
0                             None   
1                             None   
2                             None   
3                           Varied   
4                             None   

                                 Description/content  \
0  The MSc course in Molecular Medicine conveys b...   
1  Our programme offers basic and advanced lectur...   
2  Didactics of German language, literature, and ...   
3  In the 21st century's societies, simulation te...   
4  The Bonn-Cologne Graduate School of Physics an...   

                                 Types of assessment  \
0  Each of the modules of the Master's programme ...   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                    Interim reports, thesis defence   

                              Integrated internships  \
0                        Not systematically provided   
1  An internship of eight weeks is to be complete...   
2  A teaching module is an integral part of the p...   
3  The educational programme does not include a m...   
4                                                NaN   

  Course-specific, integrated German language courses  \
0                                                Yes    
1                                                 No    
2                                                 No    
3                                                 No    
4                                                 No    

                               Semester contribution  \
0  Semester fee (student services and student sel...   
1                               115 EUR per semester   
2  The university charges a registration fee of 5...   
3                       Approx. 180 EUR per semester  

In [9]:
programmes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2047 entries, 0 to 2046
Data columns (total 20 columns):
 #   Column                                               Non-Null Count  Dtype 
---  ------                                               --------------  ----- 
 0   Couse ID                                             2047 non-null   int64 
 1   University                                           2047 non-null   object
 2   Program                                              2047 non-null   object
 3   Degree                                               1705 non-null   object
 4   Teaching language                                    2002 non-null   object
 5   Languages                                            1729 non-null   object
 6   Programme duration                                   1705 non-null   object
 7   Beginning                                            1705 non-null   object
 8   Application deadline                                 1638 non-null   object
 9

It can be seen in the table above that there are 19 features - Most of them have a big ammount of not null objects (Strings). WE want to transform some of them into numbers, like teaching language, or program duration, or some tuition, or semester contribution

In [10]:
programmes

Couse ID                                         University  \
0         4000          Friedrich Schiller University Jena • Jena   
1         4001                  University of Augsburg • Augsburg   
2         4002             Philipps-Universität Marburg • Marburg   
3         4003                University of Stuttgart • Stuttgart   
4         4004                       University of Cologne • Köln   
...        ...                                                ...   
2042      3993  University of Applied Sciences Ravensburg-Wein...   
2043      3995  Otto von Guericke University Magdeburg • Magde...   
2044      3996  Dortmund University of Applied Sciences and Ar...   
2045      3997  FH Aachen University of Applied Sciences • Jülich   
2046      3998  FH Aachen University of Applied Sciences • Jülich   

                                                Program  \
0     Master of Science in Molecular Medicine MSc Mo...   
1     Master of Science in Software Engineering Mast...   
2     German as a Foreign Language (MA) German as a ...   
3     Graduate School Simulation Technology (GS SimT...   
4     Bonn-Cologne Graduate School of Physics and As...   
...                                                 ...   
2042  Electrical Engineering and Information Technol...   
2043  International Management, Marketing, Entrepren...   
2044  European Master's in Project Management (EuroM...   
2045  Bachelor of Science in Applied Chemistry Bache...   
2046  Bachelor of Engineering in Electrical Engineer...   

                                                 Degree Teaching language  \
0               Master of Science in Molecular Medicine           English   
1             Master of Science in Software Engineering           English   
2                                        Master of Arts            German   
3                                                   PhD    German English   
4                            Doctoral Degree in Physics           English   
...                                                 ...               ...   
2042                     Bachelor of Engineering (BEng)    German English   
2043  Master of Science in International Management,...           English   
2044                                Master of Arts (MA)           English   
2045           Bachelor of Science in Applied Chemistry            German   
2046  Bachelor of Engineering in Electrical Engineering            German   

                                              Languages Programme duration  \
0                          Courses are held in English.        4 semesters   
1     The courses are in English unless all students...        4 semesters   
2         German in courses and for the Master's thesis        4 semesters   
3                   English about 60%, German about 40%        6 semesters   
4            The entire programme is taught in English.        6 semesters   
...                                                 ...                ...   
2042  English (semesters 1 to 4), German (semesters ...        7 semesters   
2043  Courses and examinations are held exclusively ...        4 semesters   
2044  Courses are held in English (100%). Students c...        4 semesters   
2045  The course of study is conducted in German. A ...        6 semesters   
2046  The course of study is conducted in German. A ...        6 semesters   

                                   Beginning  \
0                            Winter semester   
1                            Winter semester   
2                            Winter semester   
3     Only for doctoral programmes: any time   
4     Only for doctoral programmes: any time   
...                                      ...   
2042                         Summer semester   
2043              Winter and summer semester   
2044                         Winter semester   
2045                         Winter semester   
2046                         Winter semester   

                             

## Data cleaning itself

- ~~Separate University from City~~
- ~~Drop all the null records from language~~
- ~~Split English / German / Others in teaching languages~~
- ~~Transform Programme duration in int32~~
- ~~Clean the begginnin and leave only the season~~
- ~~Clean the deadline~~
- ~~Make Tution a number , if it is None or simmiar put 0~~
- ~~Create a column where you can filter as masters / Phd / Bacherlor / Other~~
- ~~Make semester contribution in number~~
- Submit application leave just a link (Optional)
- ~~Delete type of assesment~~
- Integrated internships should be yes or No
- ~~Rename the columns that are wrong~~
- ~~Create a total contribution~~

### Separate University from City

In [11]:
programmes['University'].apply(lambda x: x.split('•')[0].strip()).unique()

array(['Friedrich Schiller University Jena', 'University of Augsburg',
       'Philipps-Universität Marburg', 'University of Stuttgart',
       'University of Cologne', 'Ruhr-Universität Bochum',
       'Technische Universität Berlin',
       'FH Aachen University of Applied Sciences',
       'University of Münster', 'Freie Universität Berlin',
       'University of Göttingen', 'FAU Erlangen-Nürnberg',
       'University of Passau', 'TU Bergakademie Freiberg',
       'Ludwig-Maximilians-Universität München',
       'Helmholtz Centre for Environmental Research - UFZ',
       'Leipzig University', 'Hamburg University of Applied Sciences',
       'Brandenburg University of Technology Cottbus-Senftenberg',
       'Offenburg University of Applied Sciences', 'University of Bonn',
       'Charité - Universitätsmedizin Berlin',
       'SRH University Heidelberg',
       'SRH Berlin University of Applied Sciences',
       'University of Konstanz',
       'TH Köln (University of Applied Sciences

In [12]:
programmes['City'] = programmes['University'].apply(lambda x: x.split('•')[1].strip())
programmes['University'] = programmes['University'].apply(lambda x: x.split('•')[0].strip())

In [13]:
programmes

Couse ID                                         University  \
0         4000                 Friedrich Schiller University Jena   
1         4001                             University of Augsburg   
2         4002                       Philipps-Universität Marburg   
3         4003                            University of Stuttgart   
4         4004                              University of Cologne   
...        ...                                                ...   
2042      3993  University of Applied Sciences Ravensburg-Wein...   
2043      3995             Otto von Guericke University Magdeburg   
2044      3996   Dortmund University of Applied Sciences and Arts   
2045      3997           FH Aachen University of Applied Sciences   
2046      3998           FH Aachen University of Applied Sciences   

                                                Program  \
0     Master of Science in Molecular Medicine MSc Mo...   
1     Master of Science in Software Engineering Mast...   
2     German as a Foreign Language (MA) German as a ...   
3     Graduate School Simulation Technology (GS SimT...   
4     Bonn-Cologne Graduate School of Physics and As...   
...                                                 ...   
2042  Electrical Engineering and Information Technol...   
2043  International Management, Marketing, Entrepren...   
2044  European Master's in Project Management (EuroM...   
2045  Bachelor of Science in Applied Chemistry Bache...   
2046  Bachelor of Engineering in Electrical Engineer...   

                                                 Degree Teaching language  \
0               Master of Science in Molecular Medicine           English   
1             Master of Science in Software Engineering           English   
2                                        Master of Arts            German   
3                                                   PhD    German English   
4                            Doctoral Degree in Physics           English   
...                                                 ...               ...   
2042                     Bachelor of Engineering (BEng)    German English   
2043  Master of Science in International Management,...           English   
2044                                Master of Arts (MA)           English   
2045           Bachelor of Science in Applied Chemistry            German   
2046  Bachelor of Engineering in Electrical Engineering            German   

                                              Languages Programme duration  \
0                          Courses are held in English.        4 semesters   
1     The courses are in English unless all students...        4 semesters   
2         German in courses and for the Master's thesis        4 semesters   
3                   English about 60%, German about 40%        6 semesters   
4            The entire programme is taught in English.        6 semesters   
...                                                 ...                ...   
2042  English (semesters 1 to 4), German (semesters ...        7 semesters   
2043  Courses and examinations are held exclusively ...        4 semesters   
2044  Courses are held in English (100%). Students c...        4 semesters   
2045  The course of study is conducted in German. A ...        6 semesters   
2046  The course of study is conducted in German. A ...        6 semesters   

                                   Beginning  \
0                            Winter semester   
1                            Winter semester   
2                            Winter semester   
3     Only for doctoral programmes: any time   
4     Only for doctoral programmes: any time   
...                                      ...   
2042                         Summer semester   
2043              Winter and summer semester   
2044                         Winter semester   
2045                         Winter semester   
2046                         Winter semester   

                             

### Drop all the null records from language

In [14]:
programmes.dropna(subset=['Teaching language'],how='any', inplace = True)

### Split English / German / Others in teaching languages

In [15]:
programmes['English'] = programmes['Teaching language'].apply(lambda x: 0 if re.search('English',x) == None else 1)
programmes['German'] = programmes['Teaching language'].apply(lambda x: 0 if re.search('German',x) == None else 1)
programmes['Other Language'] = programmes['Teaching language'].apply(lambda x: 0 if re.search('French|Spanish|Chinese|Italian|Other|Russian',x) == None else 1)

In [125]:
programmes.describe()

Couse ID      English       German  Other Language
count  2002.00000  2002.000000  2002.000000     2002.000000
mean   5013.24975     0.892607     0.231768        0.043956
std    1055.59843     0.309689     0.422067        0.205049
min    3589.00000     0.000000     0.000000        0.000000
25%    4143.25000     1.000000     0.000000        0.000000
50%    4748.50000     1.000000     0.000000        0.000000
75%    5658.00000     1.000000     0.000000        0.000000
max    7169.00000     1.000000     1.000000        1.000000

### Transform Programme duration in int32

In [16]:
programmes['Programme duration'].unique()

array(['4 semesters', '6 semesters', '7 semesters', '8 semesters',
       '3 semesters', '2 semesters', 'More than 9 semesters',
       '5 semesters', nan], dtype=object)

In [17]:
programmes[programmes['Programme duration'].isnull()] # We are gonna delete those programmes - They are irrelevant for our analysis

Couse ID                                   University  \
741       4887  Catholic University of Eichstätt-Ingolstadt   
742       4888  Catholic University of Eichstätt-Ingolstadt   
743       4889                              Kiel University   
744       4890                 Philipps-Universität Marburg   
745       4891                       Universität Heidelberg   
...        ...                                          ...   
1637      7092                         University of Passau   
1638      7093                                 TU Darmstadt   
1639      7094               Technische Universität Ilmenau   
1640      7095                     Freie Universität Berlin   
1661      7138         Institute of Molecular Biology (IMB)   

                                                Program Degree  \
741   Course I - Intensive German and Cultural Studi...    NaN   
742   Course II – Intensive German and German Litera...    NaN   
743   74th International Summer Course: Germany Toda...    NaN   
744   Hessen International Summer University Marburg...    NaN   
745   International Summer School of German Language...    NaN   
...                                                 ...    ...   
1637          Academic German Year Academic German Year    NaN   
1638  Hessen:International Summer University (ISU) –...    NaN   
1639  43rd International Summer Course – DSH Course ...    NaN   
1640  Sustainable and Social Business Sustainable an...    NaN   
1661  IMB International Summer School on Gene Regula...    NaN   

     Teaching language Languages Programme duration Beginning  \
741             German       NaN                NaN       NaN   
742             German       NaN                NaN       NaN   
743             German       NaN                NaN       NaN   
744     German English       NaN                NaN       NaN   
745             German       NaN                NaN       NaN   
...                ...       ...                ...       ...   
1637            German       NaN                NaN       NaN   
1638    German English       NaN                NaN       NaN   
1639            German       NaN                NaN       NaN   
1640           English       NaN                NaN       NaN   
1661           English       NaN                NaN       NaN   

     Application deadline Tuition fees per semester in EUR  ...  \
741                   NaN                              NaN  ...   
742                   NaN                              NaN  ...   
743                   NaN                              NaN  ...   
744                   NaN                              NaN  ...   
745                   NaN                              NaN  ...   
...                   ...                              ...  ...   
1637                  NaN                              NaN  ...   
1638                  NaN                              NaN  ...   
1639                  NaN                              NaN  ...   
1640                  NaN                              NaN  ...   
1661                  NaN                              NaN  ...   

     Semester contribution Academic admission requirements  \
741                    NaN                             NaN   
742                    NaN                             NaN   
743                    NaN                             NaN   
744                    NaN                             NaN   
745                    NaN                             NaN   
...                    ...                             ...   
1637                   NaN                             NaN   
1638                   NaN                             NaN   
1639                   NaN                             NaN   
1640                   NaN                             NaN   
1661                   NaN                             NaN   

                                  Language requirements  \
741                                           German A1   
742    

In [18]:
programmes[programmes['Programme duration'].isnull() & programmes['Languages'].isnull()].describe()

Couse ID     English      German  Other Language
count   297.000000  297.000000  297.000000      297.000000
mean   5493.390572    0.494949    0.555556        0.006734
std     698.707873    0.500818    0.497743        0.081922
min    4887.000000    0.000000    0.000000        0.000000
25%    4998.000000    0.000000    0.000000        0.000000
50%    5114.000000    0.000000    1.000000        0.000000
75%    6138.000000    1.000000    1.000000        0.000000
max    7138.000000    1.000000    1.000000        1.000000

In [24]:
programmes.dropna(subset=['Programme duration'],how='any', inplace = True)

In [25]:
programmes.describe()

Couse ID      English       German  Other Language
count  1705.000000  1705.000000  1705.000000     1705.000000
mean   4929.612317     0.961877     0.175367        0.050440
std    1084.674919     0.191550     0.380392        0.218915
min    3589.000000     0.000000     0.000000        0.000000
25%    4058.000000     1.000000     0.000000        0.000000
50%    4578.000000     1.000000     0.000000        0.000000
75%    5631.000000     1.000000     0.000000        0.000000
max    7169.000000     1.000000     1.000000        1.000000

In [19]:
programmes['Programme duration'].unique()

array(['4 semesters', '6 semesters', '7 semesters', '8 semesters',
       '3 semesters', '2 semesters', 'More than 9 semesters',
       '5 semesters', nan], dtype=object)

In [20]:
programmes[programmes['Programme duration']== 'More than 9 semesters'] # IT's gonna be a 9 but we need to say that its 9 or more

Couse ID                                         University  \
107       4118  Max Delbrück Center for Molecular Medicine (MD...   
266       4314                              University of Münster   
369       4441             Ludwig-Maximilians-Universität München   
413       4494                              University of Cologne   
415       4496                     Humboldt-Universität zu Berlin   
663       4782                            University of Göttingen   
1312      6280                            University of Göttingen   
1322      6293               Heinrich Heine University Düsseldorf   
1430      6885           Frankfurt School of Finance & Management   
1582      7037                             Universität Heidelberg   
1709      3626                                 University of Bonn   
1835      3767             Ludwig-Maximilians-Universität München   
1857      3790                             University of Tübingen   
1917      3855         Hochschule für Wirtschaft und Recht Berlin   
2015      3962                               Bielefeld University   

                                                Program  \
107   MDC International PhD Programme in Biomedical ...   
266   International and European Governance Internat...   
369   Graduate School of Quantitative Biosciences Mu...   
413   CEPLAS - Cluster of Excellence on Plant Scienc...   
415   Integrative Research Institute of Transformati...   
663                     Chemistry (PhD) Chemistry (PhD)   
1312  Max Planck School Matter to Life – Combined Ma...   
1322  CEPLAS - Cluster of Excellence on Plant Scienc...   
1430  Doctoral Programme in Accounting, Finance and ...   
1582  Max Planck School Matter to Life – Combined Ma...   
1709  Bonn Graduate School of Economics (BGSE) Bonn ...   
1835  Munich Graduate School of Economics (MGSE) Mun...   
1857  International Max Planck Research School for t...   
1917  Management International (German-French Progra...   
2015  Bielefeld Graduate School in History and Socio...   

                                                 Degree  \
107                                                 PhD   
266   Bachelor of Arts in International and European...   
369                          Doctoral Degree Dr rer nat   
413                 Doctoral Degree (Dr rer nat or PhD)   
415                                  Dr rer nat Dr phil   
663                                    Dr rer nat / PhD   
1312                            Master of Science / PhD   
1322                Doctoral Degree (Dr rer nat or PhD)   
1430                                         Dr rer pol   
1582                            Master of Science / PhD   
1709  Dr rer pol (PhD) with integrated MSc in Econom...   
1835                                        Dr oec publ   
1857      MSc and Doctor of Science Degree (Dr rer nat)   
1917  Three degrees: - Bachelor of Arts (HWR Berlin)...   
2015                    Dr phil in history or sociology   

          Teaching language  \
107                 English   
266   German English French   
369                 English   
413                 English   
415                 English   
663          German English   
1312                English   
1322                English   
1430                English   
1582                English   
1709                English   
1835                English   
1857                English   
1917  German English French   
2015         German English   

                                              Languages  \
107                         Working language is English   
266   Courses are held in German and French. English...   
369                 Courses are held in English (100%).   
413   The programme is conducted in English. Non-nat...   
415                                             English   
663   There is a broad variety of courses both in Ge...   
1312                       Courses are held in English.   
1322  The programme is conducted in Englis

In [26]:
programmes['Duration_in_semesters']=programmes['Programme duration'].apply(lambda x: x.split(' ')[0] if x.split(' ')[0] != 'More' else 9)

In [27]:
programmes['Duration_in_semesters'] = programmes['Duration_in_semesters'].astype('int32')

In [74]:
programmes.describe()

Couse ID      English       German  Other Language  \
count  1705.000000  1705.000000  1705.000000     1705.000000   
mean   4929.612317     0.961877     0.175367        0.050440   
std    1084.674919     0.191550     0.380392        0.218915   
min    3589.000000     0.000000     0.000000        0.000000   
25%    4058.000000     1.000000     0.000000        0.000000   
50%    4578.000000     1.000000     0.000000        0.000000   
75%    5631.000000     1.000000     0.000000        0.000000   
max    7169.000000     1.000000     1.000000        1.000000   

       Duration_in_semesters  
count            1705.000000  
mean                4.486804  
std                 1.380607  
min                 2.000000  
25%                 4.000000  
50%                 4.000000  
75%                 6.000000  
max                 9.000000

### Clean the begginnin and leave only the season

In [28]:
programmes['Beginning'].unique()

array(['Winter semester', 'Only for doctoral programmes: any time',
       'Winter and summer semester', 'Summer semester', 'Other'],
      dtype=object)

So for :
- Winter semester is gonna be winter semester and summer semester
- Only for doctoral.... is gona be Any time
- Winter and summer semester is gonna add 1 for both
- Other will remain other
- We are do it like dummies

In [29]:
programmes['Beginning_in_Winter'] = programmes['Beginning'].apply(lambda x: 0 if re.search('Winter',x) == None else 1)
programmes['Beginning_in_Summer'] = programmes['Beginning'].apply(lambda x: 0 if re.search('(s|S)ummer',x) == None else 1)
programmes['Beginning_in_Any_time'] = programmes['Beginning'].apply(lambda x: 0 if re.search('any time',x) == None else 1)
programmes['Beginning_in_Other'] = programmes['Beginning'].apply(lambda x: 0 if re.search('Other',x) == None else 1)

In [30]:
programmes.describe()

Couse ID      English       German  Other Language  \
count  1705.000000  1705.000000  1705.000000     1705.000000   
mean   4929.612317     0.961877     0.175367        0.050440   
std    1084.674919     0.191550     0.380392        0.218915   
min    3589.000000     0.000000     0.000000        0.000000   
25%    4058.000000     1.000000     0.000000        0.000000   
50%    4578.000000     1.000000     0.000000        0.000000   
75%    5631.000000     1.000000     0.000000        0.000000   
max    7169.000000     1.000000     1.000000        1.000000   

       Duration_in_semesters  Beginning_in_Winter  Beginning_in_Summer  \
count            1705.000000          1705.000000          1705.000000   
mean                4.486804             0.893842             0.328446   
std                 1.380607             0.308131             0.469786   
min                 2.000000             0.000000             0.000000   
25%                 4.000000             1.000000             0.000000   
50%                 4.000000             1.000000             0.000000   
75%                 6.000000             1.000000             1.000000   
max                 9.000000             1.000000             1.000000   

       Beginning_in_Any_time  Beginning_in_Other  
count            1705.000000         1705.000000  
mean                0.065103            0.011144  
std                 0.246779            0.105005  
min                 0.000000            0.000000  
25%                 0.000000            0.000000  
50%                 0.000000            0.000000  
75%                 0.000000            0.000000  
max                 1.000000            1.000000

### Clean the deadline

In [31]:
programmes['Application deadline'].fillna('No information', inplace= True)

In [32]:
programmes['Beginning_in_Winter'] = programmes['Beginning'].apply(lambda x: 0 if re.search('Winter',x) == None else 1)

In [33]:
# There are other more effective ways, but we are going to do ir more easy (It's not perfect , some month are wrong)
programmes['Deadline_in_January'] = programmes['Application deadline'].apply(lambda x: 0 if re.search('(J|j)anuary',x) == None else 1)
programmes['Deadline_in_February'] = programmes['Application deadline'].apply(lambda x: 0 if re.search('(F|f)ebruary',x) == None else 1)
programmes['Deadline_in_March'] = programmes['Application deadline'].apply(lambda x: 0 if re.search('(M|m)arch',x) == None else 1)
programmes['Deadline_in_April'] = programmes['Application deadline'].apply(lambda x: 0 if re.search('(A|a)pril',x) == None else 1)
programmes['Deadline_in_May'] = programmes['Application deadline'].apply(lambda x: 0 if re.search('(M|m)ay',x) == None else 1)
programmes['Deadline_in_June'] = programmes['Application deadline'].apply(lambda x: 0 if re.search('(J|j)une',x) == None else 1)
programmes['Deadline_in_July'] = programmes['Application deadline'].apply(lambda x: 0 if re.search('(J|j)uly',x) == None else 1)
programmes['Deadline_in_August'] = programmes['Application deadline'].apply(lambda x: 0 if re.search('(A|a)ugust',x) == None else 1)
programmes['Deadline_in_September'] = programmes['Application deadline'].apply(lambda x: 0 if re.search('(S|s)eptember',x) == None else 1)
programmes['Deadline_in_October'] = programmes['Application deadline'].apply(lambda x: 0 if re.search('(O|o)ctober',x) == None else 1)
programmes['Deadline_in_November'] = programmes['Application deadline'].apply(lambda x: 0 if re.search('(N|n)ovember',x) == None else 1)
programmes['Deadline_in_December'] = programmes['Application deadline'].apply(lambda x: 0 if re.search('(D|d)ecember',x) == None else 1)


In [140]:
programmes.describe()

Couse ID      English       German  Other Language  \
count  1705.000000  1705.000000  1705.000000     1705.000000   
mean   4929.612317     0.961877     0.175367        0.050440   
std    1084.674919     0.191550     0.380392        0.218915   
min    3589.000000     0.000000     0.000000        0.000000   
25%    4058.000000     1.000000     0.000000        0.000000   
50%    4578.000000     1.000000     0.000000        0.000000   
75%    5631.000000     1.000000     0.000000        0.000000   
max    7169.000000     1.000000     1.000000        1.000000   

       Duration_in_semesters  Beginning_in_Winter  Beginning_in_Summer  \
count            1705.000000          1705.000000          1705.000000   
mean                4.486804             0.893842             0.328446   
std                 1.380607             0.308131             0.469786   
min                 2.000000             0.000000             0.000000   
25%                 4.000000             1.000000             0.000000   
50%                 4.000000             1.000000             0.000000   
75%                 6.000000             1.000000             1.000000   
max                 9.000000             1.000000             1.000000   

       Beginning_in_Any_time  Beginning_in_Other  Deadline_in_January  ...  \
count            1705.000000         1705.000000          1705.000000  ...   
mean                0.065103            0.011144             0.219355  ...   
std                 0.246779            0.105005             0.413931  ...   
min                 0.000000            0.000000             0.000000  ...   
25%                 0.000000            0.000000             0.000000  ...   
50%                 0.000000            0.000000             0.000000  ...   
75%                 0.000000            0.000000             0.000000  ...   
max                 1.000000            1.000000             1.000000  ...   

       Deadline_in_March  Deadline_in_April  Deadline_in_May  \
count        1705.000000        1705.000000      1705.000000   
mean            0.159531           0.157185         0.269795   
std             0.366278           0.364082         0.443983   
min             0.000000           0.000000         0.000000   
25%             0.000000           0.000000         0.000000   
50%             0.000000           0.000000         0.000000   
75%             0.000000           0.000000         1.000000   
max             1.000000           1.000000         1.000000   

       Deadline_in_June  Deadline_in_July  Deadline_in_August  \
count       1705.000000       1705.000000         1705.000000   
mean           0.179472          0.349560            0.057478   
std            0.383860          0.476971            0.232822   
min            0.000000          0.000000            0.000000   
25%            0.000000          0.000000            0.000000   
50%            0.000000          0.000000            0.000000   
75%            0.000000          1.000000            0.000000   
max            1.000000          1.000000            1.000000   

       Deadline_in_September  Deadline_in_October  Deadline_in_November  \
count            1705.000000          1705.000000           1705.000000   
mean                0.102639             0.113783              0.114370   
std                 0.303576             0.317641              0.318353   
min                 0.000000             0.000000              0.000000   
25%                 0.000000             0.000000              0.000000   
50%                 0.000000             0.000000              0.000000   
75%                 0.000000             0.000000              0.000000   
max                 1.000000             1.000000              1.000000   

       Deadline_in_December  
count           1705.000000  
mean               0.093255  
std                0.290875  
min                0.000000  
25%                0.000000  
50%                0.000000  
75%           

### Make Tution a number 

- If None --> 0
- if Yes or Varied --> -1


In [34]:
programmes['Tuition fees per semester in EUR'].value_counts()

None      1128
1,500      107
Varied      91
Yes         68
10,000      20
          ... 
3,720        1
5,750        1
1,200        1
300          1
5,480        1
Name: Tuition fees per semester in EUR, Length: 129, dtype: int64

In [35]:
# Get rid of the ','
programmes['Tuition fees per semester in EUR']=programmes['Tuition fees per semester in EUR'].str.replace(',','', regex=True)

In [36]:

programmes['Tuition fees per semester in EUR'] = programmes['Tuition fees per semester in EUR'].apply(lambda x: 0 if x == 'None' else (-1 if ((x == 'Varied') | (x == 'Yes')) else x)).astype('int32')

### Create a column where you can filter as masters / Phd / Bacherlor / Other

- Ways to call a Master (Master / master / MSc
- Ways to call a Phd (Doctor/ Phd / Doctorade / Dr / Doctoral

In [37]:
programmes['Degree']

0                 Master of Science in Molecular Medicine
1               Master of Science in Software Engineering
2                                          Master of Arts
3                                                     PhD
4                              Doctoral Degree in Physics
                              ...                        
2042                       Bachelor of Engineering (BEng)
2043    Master of Science in International Management,...
2044                                  Master of Arts (MA)
2045             Bachelor of Science in Applied Chemistry
2046    Bachelor of Engineering in Electrical Engineering
Name: Degree, Length: 1705, dtype: object

In [38]:
programmes['Master'] = programmes['Degree'].apply(lambda x: 0 if re.search('(M|m)aster|(M|m)(S|s)(C|c)|MBA|MA|LLM|MCBL',x) == None else 1)
programmes['PhD'] = programmes['Degree'].apply(lambda x: 0 if re.search('(D|d)octo(r|ral|rade)|(P|p)(H|h)(D|d)|(D|d)(R|r)',x) == None else 1)
programmes['Bachelor'] = programmes['Degree'].apply(lambda x: 0 if re.search('(B|b)achelor|(B|b)(S|s)(C|c)|BEng|BA',x) == None else 1)


### Make semester contribution int

- If None --> 0
- If more than one value pick the highest one
- But if it's more than value and the higest one is less than 100 sum up all the values

In [39]:
programmes['Semester contribution'][2]

'The university charges a registration fee of 50 EUR as well as student union fees. State law requires all students to be members of the student union. These fees give students access to subsidised accommodation and meals. The fees also automatically include a free travel pass for public transport in most of the state of Hesse. In summer 2020, the student union dues amounted to approx. 335 EUR. https://www.uni-marburg.de/en/studying/life-at-umr/finance'

In [40]:
# Find all the number + EUR in every string for each sample
programmes['Contribution per semester']=programmes['Semester contribution'].apply(lambda x: re.findall('\d+.\d+ EUR|\d+ EUR',x))

In [41]:
programmes['Contribution per semester'][1261]

['80 EUR', '80 EUR']

In [42]:
# Get rid of EUR and also transform them to int to do that I delete what goes after '.' because it's almost irrelevant to 
# Also if it [] give the value of 0
programmes['Contribution per semester'] = programmes['Contribution per semester'].apply(lambda x: 0 if x ==[] else [int(i.split('-')[0].strip(' EUR').replace(',','.').split('.')[0]) for i in x])

In [43]:
# Grab the max value
programmes[programmes['Contribution per semester']!=0]['Contribution per semester'].apply(max)

0       230
1       115
2       335
3       180
4       290
       ... 
2042    170
2043    128
2044    310
2045    320
2046    320
Name: Contribution per semester, Length: 1528, dtype: int64

In [44]:
# Find the max element in the list if the value is over 100 if it less we sum all the elements in the list
programmes['Contribution per semester'] = programmes['Contribution per semester'].apply(lambda x: 0 if (x ==0) else (max(x) if max(x)>100 else sum(x)))

### Create a total contribution
Total fees + contribution per semester*number of semesters

In [45]:
programmes['Total contribution'] = (programmes['Contribution per semester']+ programmes['Tuition fees per semester in EUR'])*programmes['Duration_in_semesters']  

### Rename Some columns

In [46]:
programmes.rename(columns={'Program':'Programme'},inplace=True)

In [47]:
programmes.rename(columns={'Couse ID':'Course ID'},inplace=True)

In [48]:
programmes.drop('Types of assessment', axis=1, inplace=True)

In [49]:
programmes.describe()

Course ID  Tuition fees per semester in EUR      English  \
count  1705.000000                       1705.000000  1705.000000   
mean   4929.612317                       1050.983578     0.961877   
std    1084.674919                       2383.670294     0.191550   
min    3589.000000                         -1.000000     0.000000   
25%    4058.000000                          0.000000     1.000000   
50%    4578.000000                          0.000000     1.000000   
75%    5631.000000                          0.000000     1.000000   
max    7169.000000                      21750.000000     1.000000   

            German  Other Language  Duration_in_semesters  \
count  1705.000000     1705.000000            1705.000000   
mean      0.175367        0.050440               4.486804   
std       0.380392        0.218915               1.380607   
min       0.000000        0.000000               2.000000   
25%       0.000000        0.000000               4.000000   
50%       0.000000        0.000000               4.000000   
75%       0.000000        0.000000               6.000000   
max       1.000000        1.000000               9.000000   

       Beginning_in_Winter  Beginning_in_Summer  Beginning_in_Any_time  \
count          1705.000000          1705.000000            1705.000000   
mean              0.893842             0.328446               0.065103   
std               0.308131             0.469786               0.246779   
min               0.000000             0.000000               0.000000   
25%               1.000000             0.000000               0.000000   
50%               1.000000             0.000000               0.000000   
75%               1.000000             1.000000               0.000000   
max               1.000000             1.000000               1.000000   

       Beginning_in_Other  ...  Deadline_in_August  Deadline_in_September  \
count         1705.000000  ...         1705.000000            1705.000000   
mean             0.011144  ...            0.057478               0.102639   
std              0.105005  ...            0.232822               0.303576   
min              0.000000  ...            0.000000               0.000000   
25%              0.000000  ...            0.000000               0.000000   
50%              0.000000  ...            0.000000               0.000000   
75%              0.000000  ...            0.000000               0.000000   
max              1.000000  ...            1.000000               1.000000   

       Deadline_in_October  Deadline_in_November  Deadline_in_December  \
count          1705.000000           1705.000000           1705.000000   
mean              0.113783              0.114370              0.093255   
std               0.317641              0.318353              0.290875   
min               0.000000              0.000000              0.000000   
25%               0.000000              0.000000              0.000000   
50%               0.000000              0.000000              0.000000   
75%               0.000000              0.000000              0.000000   
max               1.000000              1.000000              1.000000   

            Master          PhD     Bachelor  Contribution per semester  \
count  1705.000000  1705.000000  1705.000000                1705.000000   
mean      0.747214     0.139589     0.150147                 221.537830   
std       0.434737     0.346662     0.357320                 136.910529   
min       0.000000     0.000000     0.000000                   0.000000   
25%       0.000000     0.000000     0.000000                 130.000000   
50%       1.000000     0.000000     0.000000                 230.000000   
75%       1.000000     0.000000     0.000000                 300.000000   
max       1.000000     1.000000     1.000000                 900.000000   

       Total contribution  
count         1705.000000  
mean          5340.947214  
std           9888.235706  
min             -6.00

## Create a CSV file

In [50]:
programmes.to_csv('./Database/DAAD_data_base_cleaned.csv', index= False)